java/src/main/native/src/row_conversion.cu

```diff
@@ -594,134 +594,108 @@ __global__ void copy_to_rows(const size_type num_rows, const size_type num_colum
   // This has been broken up for us in the tile_info struct, so we don't have
   // any calculation to do here, but it is important to note.
 
-  constexpr unsigned stages_count = NUM_TILES_PER_KERNEL_LOADED;
   auto group = cooperative_groups::this_thread_block();
+  auto warp = cooperative_groups::tiled_partition<NUM_THREADS_IN_WARP>(group);
   extern __shared__ int8_t shared_data[];
-  int8_t *shared[stages_count] = {shared_data, shared_data + shmem_used_per_tile};
 
 #ifdef ASYNC_MEMCPY_SUPPORTED
-  __shared__ cuda::barrier<cuda::thread_scope_block> tile_barrier[NUM_TILES_PER_KERNEL_LOADED];
+  __shared__ cuda::barrier<cuda::thread_scope_block> tile_barrier;
   if (group.thread_rank() == 0) {
-    for (int i = 0; i < NUM_TILES_PER_KERNEL_LOADED; ++i) {
-      init(&tile_barrier[i], group.size());
-    }
+    init(&tile_barrier, group.size());
   }
   group.sync();
 #endif // ASYNC_MEMCPY_SUPPORTED
 
-  auto const tiles_remaining =
-      std::min(static_cast<uint>(tile_infos.size()) - blockIdx.x * NUM_TILES_PER_KERNEL_TO_ROWS,
-               static_cast<uint>(NUM_TILES_PER_KERNEL_TO_ROWS));
+  auto const tile = tile_infos[blockIdx.x];
+  auto const num_tile_cols = tile.num_cols();
+  auto const num_tile_rows = tile.num_rows();
+  auto const tile_row_size = tile.get_shared_row_size(col_offsets, col_sizes);
+  auto const starting_column_offset = col_offsets[tile.start_col];
+
+  // to do the copy we need to do n column copies followed by m element copies OR
+  // we have to do m element copies followed by r row copies. When going from column
+  // to row it is much easier to copy by elements first otherwise we would need a running
+  // total of the column sizes for our tile, which isn't readily available. This makes it
+  // more appealing to copy element-wise from input data into shared matching the end layout
+  // and do row-based memcopies out.
+
+  // read each column across the tile
+  // each warp takes a column with each thread of a warp taking a row
+  // this is done with cooperative groups where each column is chosen
+  // by the tiled partition and each thread in that partition works on a row
+  for (int relative_col = warp.meta_group_rank(); relative_col < num_tile_cols;
```

Comment:
Just to make sure you are aware of the loop range: is `int` type enough here?

Reply:
This should be abundant. The tile is the number of rows and columns that fit into shared memory, so it is on the order of 144 depending on the size of the columns.